### PR TITLE
[MODTLR-9] Create PUT, DELETE endpoints for updating ECS TLRs

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
  Create PUT, DELETE endpoints for updating ECS TLRs

Resolves: [MODTLR-9](https://issues.folio.org/browse/MODTLR-9)

### CODE REVIEW

File: src/main/java/org/folio/controller/EcsTlrController.java [hunk 1]
```diff
@@ -36,4 +38,22 @@ public ResponseEntity<EcsTlr> postEcsTlr(EcsTlr ecsTlr) {
 
     return ResponseEntity.status(CREATED).body(ecsTlrService.post(ecsTlr));
   }
+
+  @Override
+  public ResponseEntity<Void> putEcsTlrById(UUID requestId, EcsTlr ecsTlr) {
+    log.debug("putEcsTlrById:: parameters requestId: {}", requestId);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/controller/EcsTlrController.java:44]: should we also log the value of `ecsTlr` via Supplier?

File: src/main/java/org/folio/service/EcsTlrService.java [hunk 2]
```diff
@@ -8,4 +8,6 @@
 public interface EcsTlrService {
   Optional<EcsTlr> get(UUID requestId);
   EcsTlr post(EcsTlr ecsTlr);
+  boolean put(UUID requestId, EcsTlr ecsTlr);
+  boolean delete(UUID requestId);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/EcsTlrService.java:12]: Service has no concept of POSTing and PUTing. Those are terms only make sense in context of controllers and clients. Service creates and updates objects instead. I would suggest renaming these two methods to `create` and `update` respectively.

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 3]
```diff
@@ -35,4 +35,24 @@ public EcsTlr post(EcsTlr ecsTlr) {
     return requestsMapper.mapEntityToDto(ecsTlrRepository.save(
       requestsMapper.mapDtoToEntity(ecsTlr)));
   }
+
+  @Override
+  public boolean put(UUID requestId, EcsTlr ecsTlr) {
+    log.debug("put:: requestId: {}, ecsTlr: {}", requestId, ecsTlr);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:41]: use suppliers

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 4]
```diff
@@ -35,4 +35,24 @@ public EcsTlr post(EcsTlr ecsTlr) {
...
+
+    return ecsTlrRepository.findById(requestId)
+      .map(ecsTlrEntity -> requestsMapper.mapDtoToEntity(ecsTlr))
+      .map(ecsTlrRepository::save)
+      .isPresent();
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:46]: I think we should choose an approach here and stick to it. Method `get()` returns an `Optional`, and controller converts this `Optional` into a response. We should probably do the same in your two methods. Or don't return `Optional` in any of the methods.

Also, if you fetch the request by ID only to make sure that it exists, take a look at repository's method `existsById()`.

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 5]
```diff
@@ -35,4 +35,24 @@ public EcsTlr post(EcsTlr ecsTlr) {
...
+  }
+
+  @Override
+  public boolean delete(UUID requestId) {
+    return ecsTlrRepository.findById(requestId)
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:51]: Try using `existsById()`

File: src/main/resources/swagger.api/ecs-tlr.yaml [hunk 6]
```diff
@@ -32,12 +32,41 @@ paths:
...
-          content:
-            text/plain:
-              schema:
-                type: string
-                example: Not found
+          $ref: '#/components/responses/notFoundResponse'
+        '500':
+          $ref: '#/components/responses/internalServerErrorResponse'
+    put:
+      description: Update ECS TLR by id
```
Comment (OleksandrVidinieiev) [inline src/main/resources/swagger.api/ecs-tlr.yaml:39]:
```suggestion
      description: Update ECS TLR by ID
```

File: src/main/resources/swagger.api/ecs-tlr.yaml [hunk 7]
```diff
@@ -32,12 +32,41 @@ paths:
...
+        '204':
+          description: Request successfully updated
+        '400':
+          $ref: '#/components/responses/badRequestResponse'
+        '404':
+          $ref: '#/components/responses/notFoundResponse'
+        '500':
+          $ref: '#/components/responses/internalServerErrorResponse'
+    delete:
+      description: Remove ECS TLR by id
```
Comment (OleksandrVidinieiev) [inline src/main/resources/swagger.api/ecs-tlr.yaml:59]:
```suggestion
      description: Remove ECS TLR by ID
```

File: src/main/resources/swagger.api/ecs-tlr.yaml [hunk 8]
```diff
@@ -63,6 +92,17 @@ components:
             $ref: 'schemas/EcsTlr.yaml#/EcsTlr'
     badRequestResponse:
       description: Validation errors
+      content:
+        application/json:
+          example:
+            errors:
+              - message: Request is invalid
+                code: notfound.request
```
Comment (OleksandrVidinieiev) [inline src/main/resources/swagger.api/ecs-tlr.yaml:100]: This probably fits better in 404. And message in 404 probably should not say "Request is invalid"

File: src/main/resources/swagger.api/ecs-tlr.yaml [hunk 9]
```diff
@@ -32,12 +32,41 @@ paths:
...
+          application/json:
+            schema:
+              $ref: "#/components/schemas/ecs-tlr"
+        required: true
+      parameters:
+        - $ref: '#/components/parameters/requestId'
+      responses:
+        '204':
+          description: Request successfully updated
+        '400':
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/resources/swagger.api/ecs-tlr.yaml:52]: 400 in both cases

File: src/main/java/org/folio/controller/EcsTlrController.java [hunk 10]
```diff
@@ -34,6 +37,28 @@ public ResponseEntity<EcsTlr> getEcsTlrById(UUID requestId) {
...
+    return ResponseEntity.status(CREATED).body(ecsTlrService.create(ecsTlr));
+  }
+
+  @Override
+  public ResponseEntity<Void> putEcsTlrById(UUID requestId, EcsTlr ecsTlr) {
+    log.debug("putEcsTlrById:: parameters requestId: {}, ecsTlr: {}", () -> requestId, () -> ecsTlr);
+
+    HttpStatus httpStatus = ecsTlrService.update(requestId, ecsTlr)
+      ? NO_CONTENT
+      : NOT_FOUND;
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/controller/EcsTlrController.java]: updated

File: src/main/java/org/folio/controller/EcsTlrController.java [hunk 11]
```diff
@@ -34,6 +37,22 @@ public ResponseEntity<EcsTlr> getEcsTlrById(UUID requestId) {
   public ResponseEntity<EcsTlr> postEcsTlr(EcsTlr ecsTlr) {
     log.debug("postEcsTlr:: parameters ecsTlr: {}", ecsTlr);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/controller/EcsTlrController.java:38]:
```suggestion
    log.debug("postEcsTlr:: parameters ecsTlr: {}", () -> ecsTlr);
```

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 12]
```diff
@@ -34,14 +34,36 @@ public Optional<EcsTlr> get(UUID id) {
...
+  public EcsTlr create(EcsTlr ecsTlr) {
+    log.debug("create:: parameters ecsTlr: {}", () -> ecsTlr);
     createRemoteRequest(ecsTlr, "university"); // TODO: replace with real tenantId
 
     return requestsMapper.mapEntityToDto(ecsTlrRepository.save(
       requestsMapper.mapDtoToEntity(ecsTlr)));
   }
 
+  @Override
+  public boolean update(UUID requestId, EcsTlr ecsTlr) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:46]: Do we need to explicitly pass `requestId` here? Can we get it from `ecsTlr`?

File: src/main/java/org/folio/controller/EcsTlrController.java [hunk 13]
```diff
@@ -34,6 +37,24 @@ public ResponseEntity<EcsTlr> getEcsTlrById(UUID requestId) {
...
 
-    return ResponseEntity.status(CREATED).body(ecsTlrService.post(ecsTlr));
+    return ResponseEntity.status(CREATED).body(ecsTlrService.create(ecsTlr));
+  }
+
+  @Override
+  public ResponseEntity<Void> putEcsTlrById(UUID requestId, EcsTlr ecsTlr) {
+    log.debug("putEcsTlrById:: parameters requestId: {}, ecsTlr: {}", () -> requestId, () -> ecsTlr);
+    boolean requestUpdated = ecsTlrService.update(requestId, ecsTlr);
+    HttpStatus httpStatus = requestUpdated  ? NO_CONTENT : NOT_FOUND;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/controller/EcsTlrController.java:47]:
```suggestion
    HttpStatus httpStatus = requestUpdated ? NO_CONTENT : NOT_FOUND;
```

File: src/main/java/org/folio/controller/EcsTlrController.java [hunk 14]
```diff
@@ -34,6 +37,24 @@ public ResponseEntity<EcsTlr> getEcsTlrById(UUID requestId) {
...
+
+    return ResponseEntity.status(httpStatus).build();
+  }
+
+  @Override
+  public ResponseEntity<Void> deleteEcsTlrById(UUID requestId) {
+    log.debug("deleteEcsTlrById:: parameters requestId: {}", requestId);
+    boolean requestDeleted = ecsTlrService.delete(requestId);
+    HttpStatus httpStatus = requestDeleted  ? NO_CONTENT : NOT_FOUND;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/controller/EcsTlrController.java:56]:
```suggestion
    HttpStatus httpStatus = requestDeleted ? NO_CONTENT : NOT_FOUND;
```

File: src/main/resources/swagger.api/examples/EcsTlr.sample [hunk 15]
```diff
@@ -0,0 +1,12 @@
+{
+    "id": "58e3b64b-c634-4afc-aba7-4d2a3ddef63d",
+    "instanceId": "2835ba05-fc4b-4ff2-972b-6ff7faaedd6e",
+    "requesterId": "4565999e-1994-4aab-9282-185731b4ce66",
+    "requestType": "Hold",
+    "requestLevel": "Title",
+    "requestExpirationDate": "2024-01-01T23:11:00-00:00",
```
Comment (OleksandrVidinieiev) [inline src/main/resources/swagger.api/examples/EcsTlr.sample:7]:
```suggestion
    "requestExpirationDate": "2024-01-01T23:11:00+00:00",
```

File: src/main/resources/swagger.api/examples/EcsTlr.sample [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+    "requestDate": "2024-01-01T00:12:00-00:00"
```
Comment (OleksandrVidinieiev) [inline src/main/resources/swagger.api/examples/EcsTlr.sample:8]:
```suggestion
    "requestDate": "2024-01-01T00:12:00+00:00"
```

**Review by OleksandrVidinieiev — APPROVED** [review]

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**